pkg/actions/tools/helm/release.go

```diff
@@ -16,9 +16,23 @@ type release struct {
 	AppVersion string `json:"app_version"`
 }
 
+type ReleasesOpts struct {
+	KubeContext string
+	Namespace   string
+}
+
 // ActionReleases completes releases
-func ActionReleases() carapace.Action {
-	return carapace.ActionExecCommand("helm", "list", "--output", "json")(func(output []byte) carapace.Action {
+func ActionReleases(opts ReleasesOpts) carapace.Action {
+	args := []string{"list", "--output", "json"}
```

Comment:
Should be wrapped in an `ActionCallback` as well so that is only executed when needed.

Reply:
Must it be? I don't see this function using a variable, which could change during execution of the completion.

Reply:
There are two major phases in carapace.
1. when commands are defined and completions registered
2. when the command line is parsed, and relevant completion (action) being executed

So when someone uses it directly, the code is (always) executed at program start:
```go
carapace.Gen(cmd).FlagCompletion(carapace.ActionMap{
  "release": Actionreleases(...),
})
```

It hasn't too much relevance in this case. And current usage wraps it in an `ActionCallback` anyway.
But I've adopted it as default approach to avoid stacking up executed code that isn't relevant.
Sometimes the code involves reading files and such which makes the impact more obvious.

Reply:
@rsteube thanks for the explanation!

Reply:
Flags would couple an action to a specific command. They are intended to be reusable between different commands and as [macros](https://rsteube.github.io/carapace-bin/spec/macros.html).

Reply:
Oh and it's also more obvious when looking at the macro command.

```sh
cd /tmp # context.Dir
export VAR=example # context.Env
carapace --macro bridge.CarapaceBin ...[context.Args]
```

This works for external commands and also enables remote completion (e.g. over ssh, inside a docker container,...).

Reply:
@Trojan295 things are a bit inconsisent due to the historic growth, but this is a pattern I've used for commonly used flags in a command:

Have a coupled action in an `./action` subfolder with gets passed the cobra command:
https://github.com/rsteube/carapace-bin/blob/master/completers/docker-compose_completer/cmd/action/user.go#L14

Have the non-coupled generic action in `/pkg/actions/tools/...`